pkg/resource/training_job/custom_set_update_input.go

```diff
@@ -0,0 +1,130 @@
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+// Use this file if conditions need to be updated based on the latest status
+// of training job which is not evident from API response
+
+package training_job
+
+import (
+	"errors"
+
+	ackcompare "github.com/aws-controllers-k8s/runtime/pkg/compare"
+	ackerr "github.com/aws-controllers-k8s/runtime/pkg/errors"
+	smv1alpha "github.com/aws-controllers-k8s/sagemaker-controller/apis/v1alpha1"
+	svcsdk "github.com/aws/aws-sdk-go/service/sagemaker"
+)
+
+// customSetUpdateInput modifies the input of UpdateTrainingJob.
+// Three conditions:
+// 1. Customer updates both profiler parameters: Recreate the input for profiler Rule.
+// 2. Customer only updates Profiler Config: Set the profiler rule configuration to nil to avoid validation error.
+// 3. Customer only updates Rule Configurations: Recreate the input for profiler Rule and set Profiler config to nil.
+//	  safer to do this because the "only add" behavior might reappear.
+func customSetUpdateInput(desired *resource, latest *resource, delta *ackcompare.Delta, input *svcsdk.UpdateTrainingJobInput) error {
+	if delta.DifferentAt("Spec.ProfilerConfig") && delta.DifferentAt("Spec.ProfilerRuleConfigurations") {
+		err := handleProfilerRuleConfig(desired, latest, input)
+		return err
+	}
+	if delta.DifferentAt("Spec.ProfilerConfig") && !delta.DifferentAt("Spec.ProfilerRuleConfigurations") {
+		input.SetProfilerRuleConfigurations(nil)
+		return nil
+	}
+	if delta.DifferentAt("Spec.ProfilerRuleConfigurations") && !delta.DifferentAt("Spec.ProfilerConfig") {
+		err := handleProfilerRuleConfig(desired, latest, input)
+		input.SetProfilerConfig(nil) // SM still assumes the profiler config is the same.
+		return err
+	}
+	return nil
+}
+
+// handleProfilerRuleConfig sets the input of the ProfilerRuleConfiguration so that
+// it is compatible with the sagemaker API.
+// Update training job is post operation wrt to the profiler parameters.
+// Because of this only NEW rules can be specified.
+// In this function we check to see if any new profiler configurstions have been added.
+func handleProfilerRuleConfig(desired *resource, latest *resource, input *svcsdk.UpdateTrainingJobInput) error {
+	profilerRuleDesired := desired.ko.Spec.ProfilerRuleConfigurations
+	profilerRuleLatest := latest.ko.Spec.ProfilerRuleConfigurations
+
+	if ackcompare.IsNil(profilerRuleDesired) {
+		return ackerr.NewTerminalError(errors.New("cannot remove a profiler rule."))
+	}
+	if ackcompare.IsNil(profilerRuleLatest) {
+		return nil
+	}
+	if len(profilerRuleDesired) < len(profilerRuleLatest) {
+		return ackerr.NewTerminalError(errors.New("cannot remove a profiler rule."))
+	}
+
+	ruleMap := map[string]int{}
+	profilerRuleInput := []*svcsdk.ProfilerRuleConfiguration{}
+	for _, rule := range profilerRuleLatest {
+		if ackcompare.IsNotNil(rule) && ackcompare.IsNotNil(rule.RuleConfigurationName) {
```

Comment:
why would the rule or the rule name be nil? rule name is required field

Reply:
took it out